CustomField - always use sql TEXT for serialized fields

colemanw opened the pull request:
Overview
----------------------------------------
Automatically sets serialized fields to use sql column type TEXT, to avoid length limits.

Before
----------------------------------------
When creating a multiselect/checkbox field, you have to set the "Database field length." Most users would probably think "What the heck does that mean?" Only a handful of CiviCRM/MySQL experts on the planet might know the non-self-explanatory answer is that you must enter a field length that's `>=` `the sum of the length of all options` `+` `the number of options` `+` `1`.

<img width="1836" height="636" alt="image" src="https://github.com/user-attachments/assets/38662a4f-012a-474b-9bee-48ff54f4e948" />


After
----------------------------------------
Here's an idea, let's *not* require end-users to figure that out. Just setting all serialized fields to use 'text' means there are no constraints on length. That confusing "Database field length" is no longer shown for serialized (multiselect/checkbox) fields.

colemanw committed:
CustomField - Always use 'text' column type for serialized fields

The length of serialized fields cannot be known; text is safest.

colemanw committed:
CustomField - Allow 'text' fields to be indexed

MySQL supports indices on text columns, it just requires a maxlength to be specified for the index.

colemanw commented:
@ufundo done. I lifted the restriction on indexing text fields and added a test.

ufundo commented on CRM/Core/BAO/CustomValueTable.php:
```diff
@@ -246,20 +246,8 @@ private static function create($customParams, $parentOperation = NULL) {
-          // TODO: this can still be insufficient if you try to serialize
-          // long text values. in this case a `text` field would be better
-          // but how to know? we could switch to only using varchar for serialized
-          // ints and text otherwise?
```
Couldn't we just make the default "text", but respect "varchar" if someone passes it. 

That means if you have a specific context where you know you want a varchar (because you are serialising the numbers 1,2,3,4 and can have a length 10 field) then you can do it using the API.

In general I find a bit annoying when the API just discards a value that I've passed

colemanw replied:
@ufundo hmm, I think the previous commit that created this switch statement was mistaken; `Text` is not a valid value for CustomField data_type. Neither is "varchar" so no, that wouldn't really work. The api doesn't actually permit you to set the MySQL "type" property, only the `data_type` proxy which gets converted to sql type by this function.

colemanw replied:
@ufundo acknowledged :)

IMO if you want full control over the sql type then we should do that, rather than the half-measure from #33932 why not add a new api param for the create action called `sql_type`?

colemanw commented:
Let's merge this after this week's branching.

colemanw committed:
CustomField - Add help advice about which fields to index

MySQL supports indices on text columns, but they're probably not helpful if the field is serialized.
When you search serialized fields you're usually using LIKE with wildcards, e.g. `custom_123 LIKE '%<sep>value<sep>%'` so the index cannot be used.
